diff --git a/widget/uikit/nsLookAndFeel.mm b/widget/uikit/nsLookAndFeel.mm
index 22a5c1ab905e..2c9df762bba3 100644
--- a/widget/uikit/nsLookAndFeel.mm
+++ b/widget/uikit/nsLookAndFeel.mm
@@ -334,7 +334,13 @@ static nscolor GetColorFromUIColor(UIColor* aColor) {
     aFontStyle.size = 14;
     aFontStyle.systemFont = true;
 
+// REYNARD: On iOS, use SF-based fonts
+#if defined(XP_IOS)
+    aFontName.AssignLiteral("-apple-system");
+#else
     aFontName.AssignLiteral("sans-serif");
+#endif
+
     return true;
   }
 
